### ISSUE TITLE
World-generation nástroje

### ISSUE THREAD

**Issue by dobrakmato** [pexelnet/starving#62]
## WorldGen

Na generáciu sveta sa používa nástroj _WorldGen_. 

Tento nástroj pozostáva prevažne z filtorv a brushov. Nástroj sa dá ovládať pomocou príkazov alebo pomocou GUI aplikácie slúžiacej na debugovanie a uľahčenie vývoja projektu Starving.

**Zmeny pomocou tohoto nástroja sú nevratné! Teda neexistuje žiadne undo.**

---
### Brush

**Brush** (štetec) je nástroj, ktorý slúži na aplikovanie filtra danú plochu. Každý štetec ma nejaký tvar a nejakú veľkosť.

**Dostupné tvary:** `square` (štvorec), `circle` (kruh)
**Dostupné veľkosti:** každé prirodzené číslo *(veľké veľkosti zaťažujú server a môžu spôsobiť jeho pád) 

> Príkaz na vybranie tvaru štetca: `/bt <tvar>`
> Príkaz na zvolenie veľkosti štetca: `/bs <velkost>`

---
### Filter

**Filter** je definícia transformácie prostredia. Na aplikovanie filtru sa používa right-click _Blaze Rod_-om, ktorý aplikuje práve používaný filter na oblasť, definovanú aktuálne používaným štetcom. Okrem toho sa dá filter aplikovať aj pomocou nástroja _WorldEdit_.

Každý filter môže mať vlastnosti / nastavenia, ktoré určujú jeho správanie (šanca na pokrytie trávou,...). Takéto vlastnosti sú troch typov:
- **float** - percentuálna hodnota uvedená v tvare desatinného čísla (`0.2` = 20%)
- **boolean** - logická hodnota _zvyčajne používaná na povolenie/zakázanie príznaku_ (`true` alebo `false`)
- **integer** - celočíslená hodnota _zvyčajne používaná na vyjadrenie počtu_ (`5`, `25`,...)

**Dostupné filtre:** GrassFilter, FieldFilter

> Príkaz na výber filtra: `/f <filter>`
> Príkaz na zobrazenie dostupných filtrov: `/f`
> Príkaz na zobrazenie dostupných vlastností filtra: `/fp`
> Príkaz na nastavenie hodnoty vlastnosti filtra: `/fp <vlanost> <hodnota>`
#### GrassFilter

Tento filter sadí _LONG_GRASS_ a _DOUBLE_PLANT_ bloky podľa vlastností / nastavení filtra.

![](http://i.imgur.com/detIcUL.png)
#### FieldFilter

Tento filter generuje _FARMLAND_ a na ňom rôzne sadenice.

![](http://i.imgur.com/SaGsxoc.png)

---
### WorldEdit

Filtre sa dajú aplikovať aj na region nástroja _WorldEdit_. Stačí vybrať dva body (_drevenou sekerou_ alebo pomocou príkazov `//pos1`, `//pos2`) a použíť príkaz na aplikovanie filtra na danú vybranú oblasť. 

> Príkaz na použitie filtra na aktuálnu oblasť: `/ar`

---
### Aplikácia

Aplikácia ponúka pekné prostredie na prácu s filtrami. Obsahuje dropdown na výber filtra. Ovládacie prvky na zmenu vlastností štetca, ovládacie prvky na zmenu vlasntostí filtra a ešte veľa ďalších vecí ako editor cinematicov, npc, skriptov...

**Aplikácia vyžaduje JRE 1.8 alebo novšie.**

Na stiahnutie je k dispozící tu: http://www.starving.eu/controller2.zip

![](http://i.imgur.com/vGh79Zg.png)

Po prvom spustení musí používateľ získať svoj _prístupový kľúč_ zo servera, na ktorom plánuje aplikáciu používať. Tento kľúč získa zadaním príkazu `/genkey` v hre _Minecraft_. Tento kód musí následne opísať / skopírovať do nastavení aplikácie, nastavenia uložiť a pokúsiť sa zvonu pripojiť (reconnectnúť) na server.

![](http://i.imgur.com/2li30xO.png)

![](http://i.imgur.com/XCVj3Nq.png)

![](http://i.imgur.com/tvCsx9F.png)

Kód sa po vygenerovaní nebude dať znovu zobraziť. Pri opätovnom zadaní príkazu sa vygeneruje nový kód. K strate kódu môže dojsť ak dojde k neúspešnému odhláseniu v tej istej relácií Minecraftu, v ktorej bol kód vygenerovaný.

Ak je kód zadaný pre zadané meno správny, dôjde k overeniu údajov a aplikácia sprístupní svoje funkcie.

![](http://i.imgur.com/utkGxA3.png)

Počas celého používania aplikácie musí byť hráč online na serveri. Po odpojení aplikácie od serveru (napríklad kvôli prerušeniu internetového pripojenia) aplikácia pokúšať znovu pripojiť automaticky.

**Comment by janscz** [comment]
Jo to je fajn, to zatravňování už ale je ne ?

**Comment by dobrakmato** [comment]
Nie, nemame zatravnovanie, to stare je nekompaktibilne, takze ho bude treba prerobit. 

Otazka je, ako by bolo idealne tieto nastroje pouzivat. Lebo niektore sa pravdepodobne budu dat aplikovat na region (sposob we), niektore by mali iba vygenerovat strukturu po kliknuti na blok a niektorymi by sa malo dat kreslit (sposob voxelpaintru). 

Okrem toho by mali mat niektore nastroje vlastnosti (napr. hustota travy, ci sadit vysoku alebo nizku,...).

Momentalne nemam nejaky sposob ako vsetky tieto vlastnosti, ktore do tych nastrojov potrebujeme dat spojit do jednotneho systemu aby nebolo nutne pre kazdy nastroj znovu definovat, ako sa bude pouzivat.

**Comment by dobrakmato** [comment]
Este mi napada jedno sialene riesenie. Pouzit prikazy na strane servera. A na strane klienta vytvorit UI aplikaciu, ktora by pisala do hry prikazy automaticky a tak zjednodusila pouzivanie tychto nastrojov.

https://batchloaf.wordpress.com/2013/02/13/sending-a-key-to-a-window-thats-not-in-focus-c-program/

**Comment by janscz** [comment]
Pokud to je efektivní řešení a bude to fungovat tak to klidně můžeme použít

**Comment by dobrakmato** [comment]
**Veľký update hlavného príspevku.**

@pitkes22 @janscz @kubob

**Comment by janscz** [comment]
tyto nástroje nám rozhodně ulehčí práci a hlavně zefektivní.

Velký pokrok ve Starvingu !
